numba/np/ufunc/deviceufunc.py

```diff
@@ -445,31 +434,27 @@ def __init__(self, func, sig, identity=None, cache=False, targetoptions={}):
         self.identity = parse_identity(identity)
         self.signature = sig
         self.inputsig, self.outputsig = parse_signature(self.signature)
-        assert len(self.outputsig) == 1, "only support 1 output"
-        # { arg_dtype: (return_dtype), cudakernel }
+
+        # Maps from a tuple of input_dtypes to (output_dtypes, cudakernel)
         self.kernelmap = OrderedDict()
 
     @property
     def pyfunc(self):
         return self.py_func
 
-    def add(self, sig=None, argtypes=None, restype=None):
-        # Handle argtypes
-        if argtypes is not None:
-            warnings.warn("Keyword argument argtypes is deprecated",
-                          DeprecationWarning)
-            assert sig is None
-            if restype is None:
-                sig = tuple(argtypes)
-            else:
-                sig = restype(*argtypes)
-        del argtypes
-        del restype
-
+    def add(self, sig=None):
```

Comment:
I don't think so - this method isn't part of the public interface - it's called [in the decorator](https://github.com/numba/numba/blob/main/numba/np/ufunc/decorators.py#L194) as `guvec.add()` with neither `argtypes` nor `restype` and then after that the opportunity to call it is removed because the decorator returns the result of `guvec.build_ufunc()`.

Reply:
Thanks for checking, I've also taken a look through Numba's internal use, it certainly matches the notion of implementation detail. It's also not advertised in the docs as an API.